File: LeetCodeExample.Test/Binary Tree/_00222_CountCompleteTreeNodes.cs
using LeetCodeExample.Test.Common;
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Linq;

namespace LeetCodeExample.Test;

/// <summary>
//Given the root of a complete binary tree, return the number of the nodes in the tree.
//According to Wikipedia, every level, except possibly the last, is completely filled in a complete binary tree, and all nodes in the last level are as far left as possible.It can have between 1 and 2h nodes inclusive at the last level h.
//Design an algorithm that runs in less than O(n) time complexity.
/// </summary>
public class _00222_CountCompleteTreeNodes
{
    public int CountNodes(TreeNode root)
    {
        if (root == null)
            return 0;
        int height = GetHeight(root);
        return GetCount(root, 1, 1, height);
    }

    int GetHeight(TreeNode node)
    {
        // The left side of a complete binary tree will always have the height.
        // This is O(h).
        if (node == null)
            return 0;
        return GetHeight(node.left) + 1;
    }

    int GetCount(TreeNode node, int c, int h, int target_height)
    {
        // base case
        if (node.left == null && node.right == null)
        {
            if (h == target_height)
            {
                // Bingo.
                //Console.WriteLine($"c {c}, h {h}");
                int count = 0;
                for (int i = 0; i < h - 1; i++)
                {
                    count += (int)Math.Pow(2, i);
                }
                return count + c;
                //return (int)Math.Pow(2, h - 1) / 2 + c;
                //return 2 * (h - 1) + c;
            }
            return -1;
        }

        // Opportunistically traverse down the right side to locate 
        // the last node.
        if (node.right != null)
        {
            int rCount = GetCount(node.right, c * 2, h + 1, target_height);
            // If the right path does not lead to the target height we 
            // need to check the left path.
            if (rCount != -1)
            {
                return rCount;
            }
        }
        if (node.left != null)
        {
            return GetCount(node.left, c * 2 - 1, h + 1, target_height);
        }
        // If neither paths reach the target height then we need to try 
        // another route from above.
        return -1;
    }
}

File: LeetCodeExample.Test/Binary Tree/_00222_CountCompleteTreeNodes.py
# Given the root of a complete binary tree, return the number of the nodes in the tree.

# According to Wikipedia, every level, except possibly the last, is completely filled in a complete binary tree, and all nodes in the last level are as far left as possible. It can have between 1 and 2h nodes inclusive at the last level h.

# Design an algorithm that runs in less than O(n) time complexity.

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        return 1 + self.countNodes(root.right) + self.countNodes(root.left) if root else 0
    


class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        self.maxHeight = 0
        self.ans = 0
        self.isDone = False
        def rec(node: Optional[TreeNode], h: int, val: int):
            if not node:
                return 0
            if self.isDone:
                return 0
            # Traverse down the right side until we get to that level
            self.maxHeight = max(self.maxHeight, h)
            self.ans = max(self.ans, val)
            
            rec(node.left, h + 1, val * 2)
            if node.left and not node.right:
                self.isDone = True
                return
            rec(node.right, h + 1, val * 2 + 1)

        rec(root, 1, 1)
        return self.ans